torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1546,6 +1607,24 @@ def _prep_grads_for_backward(self) -> None:
                 p.grad.size() != p._orig_size  # type: ignore[attr-defined]
                 or p.grad.device != p.device
             ):
+                offloaded: bool = p.grad.device != p.device
+                if offloaded:
+                    assert self.cpu_offload.offload_params, \
+                        "`p.grad.device` and `p.device` should be the same " \
+                        "if not offloading parameters to CPU"
+                outside_no_sync: bool = p.grad.size() == p._local_shard.shape  # type: ignore[attr-defined]
+                assert not outside_no_sync or \
+                    (self._is_root or self._require_backward_grad_sync), \
+                    "Each FSDP modules should require backward grad sync " \
+                    "outside `no_sync()` except the root module"
```

Comment:
curious why except the root module?

Reply:
This is a great catch. I conflated two conditions together incorrectly.

I have a clearer version now.